share/wake/lib/core/lens.wake

```diff
@@ -0,0 +1,156 @@
+# Copyright 2021 SiFive, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You should have received a copy of LICENSE.Apache2 along with
+# this software. If not, you may obtain a copy at
+#
+#    https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+package wake
+
+export def (l: List a) |> (t: a => List b): List b = mapFlat t l
+
+# TODO: How do we feel about `x | f |> g` vs `|> x |> f |> g`?
+export def |> (x: a): List a = x, Nil
+
+# NOTE: These are not part of an offline discussion but I think they're a
+#       useful concept. They're a kind of horizontial composition rather
+#       than vertical. You can use them to do things like set/get a value
+#       in a fallback like way. We can drop these. getJString made me
+#       think we might want something like this for some unknown reason
+#
+# Example:
+#     def getJString j =
+#         |> j
+#         |> jString ++> (_ | jArray |> jString)
```

Comment:
Hmm. While I understand that the `++>` binds the `jString` and anonymous function, I think that's not obvious to most readers. Recognizing precedence with unfamiliar operators is hard. I think it would be a better style to give the two halves a name and then start a new pipeline with `(name1 ++ name2) |> ...`.

Reply:
OTOH, if you had used explicit parens, maybe it would've been ok:
```
def petNames =
  |> json
  |> getField "pets"
  |> (getField "cat" ++> getField "dog")
  |> getField "name"
  |> getJString
  | catWith ", "
```

Reply:
OTOOH, you could almost as easily (and definitely more concretely) have written:
```
def petNames =
  |> json
  |> getField "pets"
  |> (\x getField "cat" x ++ getField "dog" x)
  |> getField "name"
  |> getJString
  | catWith ", "
```